rfd/0153-resource-guidelines.md

```diff
@@ -0,0 +1,610 @@
+---
+authors: Tim Ross (tim.ross@goteleport.com)
+state: draft
+---
+
+# RFD 153 - Resource Guidelines
+
+## Required approvers
+
+- Engineering: `@zmb3`
+
+## What
+
+Guidelines for creating backend resources and APIs to interact with them.
+
+## Why
+
+To date resources follow slightly different patterns and have inconsistent APIs for creating, updating and deleting
+them, see [#29234](https://github.com/gravitational/teleport/issues/29234). In addition to reducing confusion and
+fostering a better developer experience, this will also remove complexity from the terraform provider and teleport
+operator that have to accommodate for every subtle API and resource difference that exist today.
+
+## Details
+
+### Defining a resource
+
+A resource at minimum MUST include a kind, version, `teleport.header.v1.Metadata` and a specification message with any
+additional parameters needed to represent the resource. While the kind and version may seem like they would be easy to
+derive from the message definition itself, they need to be defined so that anything processing a generic resource can
+identify which resource is being processed. For example, `tctl` interacts with resources in their in raw yaml text form
+and leverage `services.UnknownResource` to identify the resource and act appropriately.
+
+```protobuf
+import "teleport/header/v1/resourceheader.proto";
+
+// Foo is a resource that does foo.
+message Foo {
+  // The kind of resource represented.
+  string kind = 1;
+  // An optional subkind to differentiate variations of the same kind.
+  string sub_kind = 2;
+  // The version of the resource being represented.
```

Comment:
It's a good question. I do think an api_version can be useful, as it can be a indicator that certain features are supported by the backend, but I also suspect the main reason it exists in projects like k8s is because YAML has a harder time representing that. We've relied on versions before to toggle different behavior (think roles); whether that's a good idea is debatable, but it adds design flexibility.

I think we can "solve" some of my gripes with documentation, stating clearly what "version" and what "revision" mean, and how one is not the other (as in many other places "version" is used for what we call "revision"). (I would prefer to call it api_version and be done with it, though, but it doesn't have to happen here.)

Reply:
Also adding to this, I'd rather never see FooV2 _or_ foo.v2 - API major bumps are a pain and should be a very rare exception to the rule.